Mathlib/NumberTheory/ArithmeticFunction/Carmichael.lean

```diff
@@ -0,0 +1,165 @@
+/-
+Copyright (c) 2025 Snir Broshi. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Snir Broshi
+-/
+module
+
+public import Mathlib.Algebra.GCDMonoid.FinsetLemmas
+public import Mathlib.NumberTheory.ArithmeticFunction.Defs
+public import Mathlib.RingTheory.ZMod.UnitsCyclic
+
+/-!
+# The Carmichael function
+
+## Main definitions
+
+* `ArithmeticFunction.Carmichael`: the Carmichael function `λ`,
+  also known as the reduced totient function.
+
+## Main results
+
+* A formula for `λ n` in terms of the prime factorization of `n`, given by the following theorems:
+  `carmichael_two_pow_of_le_two`, `carmichael_two_pow_of_ne_two`, `carmichael_pow_of_prime_ne_two`,
+  and `carmichael_factorization`.
+
+## Notation
+
+We use the standard notation `λ` to represent the Carmichael function,
+which is accessible in the scope `ArithmeticFunction.Carmichael`.
+Since the notation conflicts with the anonymous function notation, it is impossible to use this
+notation in statements, but the pretty-printer will use it when showing the goal state.
+
+## Tags
+
+arithmetic functions, totient
+-/
+
+@[expose] public section
+
+open Nat Monoid
+
+variable {R : Type*}
+
+namespace ArithmeticFunction
+
+/-- `λ` is the Carmichael function, also known as the reduced totient function,
+defined as the exponent of the unit group of `ZMod n`. -/
+def Carmichael : ArithmeticFunction ℕ where
+  toFun
+    | 0 => 0
+    | n + 1 => Nat.find <| ExponentExists.of_finite (G := (ZMod (n + 1))ˣ)
+  map_zero' := rfl
+
+@[inherit_doc]
+scoped[ArithmeticFunction.Carmichael] notation "λ" => ArithmeticFunction.Carmichael
+
+open scoped Carmichael
+
+theorem carmichael_eq_exponent {n : ℕ} (hn : n ≠ 0) : Carmichael n = exponent (ZMod n)ˣ := by
+  cases n with | zero => contradiction | succ n =>
+  change Nat.find _ = _
+  grind [exponent, ExponentExists.of_finite]
+
+/-- This takes in an `NeZero n` instance instead of an `n ≠ 0` hypothesis. -/
+theorem carmichael_eq_exponent' (n : ℕ) [NeZero n] : Carmichael n = exponent (ZMod n)ˣ :=
+  carmichael_eq_exponent <| NeZero.ne n
+
+@[simp]
+theorem pow_carmichael {n : ℕ} (a : (ZMod n)ˣ) : a ^ Carmichael n = 1 := by
+  cases n
+  · rw [map_zero, pow_zero]
+  rw [carmichael_eq_exponent']
+  exact pow_exponent_eq_one a
+
+theorem carmichael_dvd_totient (n : ℕ) : Carmichael n ∣ n.totient := by
+  by_cases hn : n = 0
+  · simp [hn]
+  have : NeZero n := ⟨hn⟩
+  rw [← ZMod.card_units_eq_totient, carmichael_eq_exponent']
+  exact Group.exponent_dvd_card
+
+theorem carmichael_dvd {a b : ℕ} (h : a ∣ b) : Carmichael a ∣ Carmichael b := by
+  by_cases hb : b = 0
+  · simp [hb]
+  rw [carmichael_eq_exponent <| ne_zero_of_dvd_ne_zero hb h, carmichael_eq_exponent hb]
+  have : NeZero b := ⟨hb⟩
```

Comment:
```suggestion
  cases b
  · simp
  have : NeZero a := ⟨ne_zero_of_dvd_ne_zero (by grind) h⟩
  simp_rw [carmichael_eq_exponent']
```
Not sure that this is better, just an alternative (setting up `NeZero a` feels a bit ugly...).